### ISSUE TITLE
Compile problems.

### ISSUE THREAD

**Issue by lupohirp** [flutter/flutter#76456]
Hi all,

i've a big problem with Crostini and Flutter on Chrome OS.

When i try to build my flutter application, for the first time it goes all ok, but suddenly, if i try to rebuild, it stucks on assembleDebug. Could be a memory issue, but i have 8GB Ram and 16GB Swap partition. And also i,ve tried with Cloudready with unfortunately the same result. On Chrome OS 85/86 didn't happened.

I'm using latest Flutter stable(1.22.6).

Here the situation when it stucks :

![image](https://user-images.githubusercontent.com/1065547/108599653-bee48980-7392-11eb-989f-bc30412eddd0.png)
I don't think that it could be a memory problem.
Do you have any solution?

**Comment by iapicca** [comment]
@lupohirp 
> i've a big problem with Crostini and Flutter on Chrome OS.

you have all my sympathy 

can you provide your `flutter doctor -v`and  your `flutter build <whatever_you_are_building_fo> -v` 
please avoid posting picture when simple text would do, 
if the output is too lengthy  consider attaching it as `.txt` file

**Comment by lupohirp** [comment]
> @lupohirp
> 
> > i've a big problem with Crostini and Flutter on Chrome OS.
> 
> you have all my sympathy
> 
> can you provide your `flutter doctor -v`and your `flutter build <whatever_you_are_building_for> -v`
> please avoid posting picture when simple text would do,
> if the output is too lengthy consider attaching it as `.txt` file

yeah sure. 

this is my flutter doctor -v: 
`✓] Flutter (Channel stable, 1.22.6, on Linux, locale en_US.UTF-8)
    • Flutter version 1.22.6 at /home/pakilodi/flutter
    • Framework revision 9b2d32b605 (4 weeks ago), 2021-01-22 14:36:39 -0800
    • Engine revision 2f0af37152
    • Dart version 2.10.5

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /home/pakilodi/Android/Sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[!] Android Studio
    • Android Studio at /opt/android-studio
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (1 available)
    • emulator-5554 (mobile) • emulator-5554 • android-x64 • Android 9 (API 28)
`
and attached my build when it goes to ko. Seems a memory issue, but as you can see, i have 16 gb free of swap partition. 
[build_ko.txt](https://github.com/flutter/flutter/files/6020862/build_ko.txt)

**Comment by darshankawar** [comment]
@lupohirp 
Does it get stuck everytime you try to rebuild ? Also, by rebuilding, you mean `hot restart`, `hot reload` or simply `flutter run` ?

**Comment by lupohirp** [comment]
> @lupohirp
> Does it get stuck everytime you try to rebuild ? Also, by rebuilding, you mean `hot restart`, `hot reload` or simply `flutter run` ?

nope, i mean `flutter run`. Infact, after the first succesful build,` hot restart` and` hot reload` works without problems. But if i do another `flutter run` ( or another gradle operation like a `flutter build apk` for example ).  At first sight seems a memory leak problem, but i'm not sure of this. Don't know honestly if it is a Crostini problem or Flutter problem.....

**Comment by darshankawar** [comment]
Thanks. I am not much aware about Crostini and Chromebook setup and don't have machine to verify at my end to identify root-cause. Keeping this open for further investigation and triage.

**Comment by jmagman** [comment]
Last lines of logs are
```
[+31342 ms] > Task :firebase_admob:compileDebugJavaWithJavac
[+11138 ms] > Task :firebase_admob:extractDebugAnnotations
[ +704 ms] > Task :firebase_admob:mergeDebugGeneratedProguardFiles UP-TO-DATE
[ +482 ms] > Task :firebase_admob:mergeDebugConsumerProguardFiles UP-TO-DATE
[+3804 ms] > Task :firebase_admob:syncDebugLibJars
[+6849 ms] > Task :firebase_admob:bundleDebugAar
[+1220 ms] > Task :firebase_admob:bundleLibRuntimeToJarDebug
[ +385 ms] > Task :firebase_admob:compileDebugSources
[+1007 ms] > Task :firebase_admob:bundleLibCompileToJarDebug
[ +305 ms] > Task :firebase_admob:assembleDebug
```

**Comment by Hixie** [comment]
It's like it's hanging during the gradle step where we compile the code, weird.

**Comment by Hixie** [comment]
Does this happen for any application, or just yours? e.g. does it reproduce with the template "flutter create" app?

**Comment by lupohirp** [comment]
> Does this happen for any application, or just yours? e.g. does it reproduce with the template "flutter create" app?

it happens also on a vanilla project generated by flutter create. it compiles well the first time and then stuck.

**Comment by lupohirp** [comment]
> Last lines of logs are
> 
> ```
> [+31342 ms] > Task :firebase_admob:compileDebugJavaWithJavac
> [+11138 ms] > Task :firebase_admob:extractDebugAnnotations
> [ +704 ms] > Task :firebase_admob:mergeDebugGeneratedProguardFiles UP-TO-DATE
> [ +482 ms] > Task :firebase_admob:mergeDebugConsumerProguardFiles UP-TO-DATE
> [+3804 ms] > Task :firebase_admob:syncDebugLibJars
> [+6849 ms] > Task :firebase_admob:bundleDebugAar
> [+1220 ms] > Task :firebase_admob:bundleLibRuntimeToJarDebug
> [ +385 ms] > Task :firebase_admob:compileDebugSources
> [+1007 ms] > Task :firebase_admob:bundleLibCompileToJarDebug
> [ +305 ms] > Task :firebase_admob:assembleDebug
> ```

this log is captured with the latest gradle distribution, but it happens also with the one shipped with flutter(5.6.2 if i don't get wrong)

**Comment by Hixie** [comment]
It would be great if we could confirm that this happens for other people as well on ChromeOS/Crostini. That would help us determine if it's a general ChromeOS bug or a specific issue with @lupohirp's setup.

**Comment by lupohirp** [comment]
> It would be great if we could confirm that this happens for other people as well on ChromeOS/Crostini. That would help us determine if it's a general ChromeOS bug or a specific issue with @lupohirp's setup.

if you want, i can attach my pubspec setup. Maybe that's something that is wrong....i've opened also a ticket to crbug, because i don't know if it is a problem of crostini or flutter.

**Comment by jpnurmi** [comment]
I'm not sure how to create emulators on Chrome OS but works fine for me with a real device.

- Chromebook Pixel (2015)
- Chrome OS Version 88.0.4324.186 (Official Build) (64-bit)
- Pixel 5
- Android 11 (RQ1A.210205.004)

<details><summary>flutter doctor -v</summary>

```
jpnurmi@penguin:~$ flutter doctor -v
[✓] Flutter (Channel stable, 1.22.6, on Linux, locale en_US.UTF-8)
    • Flutter version 1.22.6 at /home/jpnurmi/Flutter
    • Framework revision 9b2d32b605 (6 weeks ago), 2021-01-22 14:36:39 -0800
    • Engine revision 2f0af37152
    • Dart version 2.10.5

 
[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0-rc1)
    • Android SDK at /home/jpnurmi/Android/Sdk
    • Platform android-30, build-tools 31.0.0-rc1
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[!] Android Studio
    • Android Studio at /opt/android-studio
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (1 available)
    • Pixel 5 (mobile) • 0B261FDD40007Y • android-arm64 • Android 11 (API 30)

! Doctor found issues in 1 category.
```
</details>

**Comment by jpnurmi** [comment]
Perhaps running `strace -f flutter run` to see what is happening under the hood helps to figure out what's going on.

**Comment by lupohirp** [comment]
> Perhaps running `strace -f flutter run` to see what is happening under the hood helps to figure out what's going on.

it remains blocked here: 

```
[pid  1971] <... futex resumed> )       = 0
[pid  2104] <... close resumed> )       = 0
[pid  1971] futex(0x7d31287ddd28, FUTEX_WAIT_PRIVATE, 2, NULL <unfinished ...>
[pid  2104] futex(0x7d3129e6cb7c, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid  1930] <... futex resumed> )       = 1
[pid  2108] <... futex resumed> )       = 0
[pid  2104] <... futex resumed> )       = 1
[pid  2108] futex(0x7d3129e6cb28, FUTEX_WAIT_PRIVATE, 2, NULL <unfinished ...>
[pid  2104] futex(0x7d3129e6cb28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid  2108] <... futex resumed> )       = -1 EAGAIN (Resource temporarily unavailable)
[pid  2104] <... futex resumed> )       = 0
[pid  2108] futex(0x7d3129e6cb28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid  1930] futex(0x7d31287ddd28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid  2108] <... futex resumed> )       = 0
[pid  2155] getcwd( <unfinished ...>
[pid  1971] <... futex resumed> )       = 0
[pid  2155] <... getcwd resumed> "/home/pakilodi/StudioProjects/buondeal_business", 4096) = 48
[pid  2108] stat("/home/pakilodi/Android/Sdk/platforms/android-28/android.jar",  <unfinished ...>
[pid  2104] stat("/home/pakilodi/Android/Sdk/build-tools/29.0.2/core-lambda-stubs.jar",  <unfinished ...>
[pid  1971] futex(0x7d31287ddd28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid  1930] <... futex resumed> )       = 1
[pid  2108] <... stat resumed> {st_mode=S_IFREG|0644, st_size=46784255, ...}) = 0
[pid  2104] <... stat resumed> {st_mode=S_IFREG|0644, st_size=18159, ...}) = 0
[pid  1971] <... futex resumed> )       = 0
[pid  2108] futex(0x7d3129e6e778, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid  1928] <... futex resumed> )       = 0
[pid  2109] <... futex resumed> )       = 0
[pid  2155] rt_sigprocmask(SIG_BLOCK, [PROF],  <unfinished ...>
[pid  2109] futex(0x7d3129e6e728, FUTEX_WAIT_PRIVATE, 2, NULL <unfinished ...>
[pid  2155] <... rt_sigprocmask resumed> [QUIT TTOU], 8) = 0
[pid  2108] <... futex resumed> )       = 1
[pid  2155] newfstatat(AT_FDCWD, "/home/pakilodi/.pub-cache/hosted/pub.dartlang.org/im_stepper-0.1.2+10/lib/src/dot_stepper/core/dot_offset.dart",  <unfinished ...>
[pid  2108] futex(0x7d3129e6e728, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid  1971] futex(0x7d3129da3178, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid  2109] <... futex resumed> )       = 0
[pid  2108] <... futex resumed> )       = 1
[pid  2155] <... newfstatat resumed> {st_mode=S_IFREG|0644, st_size=768, ...}, 0) = 0
[pid  2109] futex(0x7d3129e6e728, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid  2155] rt_sigprocmask(SIG_SETMASK, [QUIT TTOU],  <unfinished ...>
[pid  2109] <... futex resumed> )       = 0
[pid  2155] <... rt_sigprocmask resumed> NULL, 8) = 0
[pid  2108] futex(0x7d31a0080b7c, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid  2155] rt_sigprocmask(SIG_BLOCK, [PROF],  <unfinished ...>
```

could be the im_stepper dependency? but it happens also on a vanilla project....

**Comment by Hixie** [comment]
@jpnurmi I wasn't quite sure I understood your comments -- are you saying you can reproduce this issue or cannot reproduce this issue?

**Comment by jpnurmi** [comment]
@Hixie I was not able to reproduce the issue.

**Comment by lupohirp** [comment]
@jpnurmi  what are your specs? Mine are these:

Core I5 8250U, 8GB RAM

**Comment by jpnurmi** [comment]
I listed the specs above: https://github.com/flutter/flutter/issues/76456#issuecomment-789710727

**Comment by lupohirp** [comment]
> I listed the specs above: [#76456 (comment)](https://github.com/flutter/flutter/issues/76456#issuecomment-789710727)

so you have 16GB of RAM?

**Comment by jpnurmi** [comment]
> so you have 16GB of RAM?

Yeah

![image](https://user-images.githubusercontent.com/140617/110616814-456fd800-8195-11eb-89de-9f3f7bf8446a.png)

**Comment by Hixie** [comment]
If we're unable to reproduce the problem it's going to be very difficult to debug it.

@lupohirp Any chance you have access to some other devices and can see if you can reproduce it on them too? I'm curious what the pattern is for when it happens and when it doesn't.

**Comment by lupohirp** [comment]
> If we're unable to reproduce the problem it's going to be very difficult to debug it.
> 
> @lupohirp Any chance you have access to some other devices and can see if you can reproduce it on them too? I'm curious what the pattern is for when it happens and when it doesn't.

not at all. i'm sure that is a memory problem. Infact on Crostini i have 6.6 gb of ram and 3 free. Are not much yeah,  but i noticed the same strange thing during compile: it stays on 1 gb free, and suddenly it goes to zero. Here Crostini hit OOM. But sometimes it goes well. But free ram isn't the same when the compilation started, but under the 1 gb. Seems a memory leak problem?

**Comment by lupohirp** [comment]
i noticed one thing(that it is responsible for OOM) : why there are all these java processes after compilation? Are flutter supposed to kill them after he compiles the code? or i'm wrong? 

![image](https://user-images.githubusercontent.com/1065547/112010885-02641c00-8b28-11eb-8da8-9ed6c9fc645f.png)

**Comment by Hixie** [comment]
We should try to find another 8GB ChromeOS machine and see if it reproduces the problem.

**Comment by lupohirp** [comment]
> @lupohirp Any chance you have access to some other devices and can see if you can reproduce it on them too? I'm curious what the pattern is for when it happens and when it doesn't.
sorry i missed the second part of the question. I have an another machine with same specs(8GB RAM), but with windows, and works. My memory stays on 90%, but i can work. And we are talking of the same project. Crostini hit out of memory, it's clear. But previously didn't happened.

**Comment by Hixie** [comment]
One thing to look at is maybe lowering the heap used by Gradle. Lowering it too much will cause a crash but you may be able to do it. Look for `org.gradle.jvmargs=-Xmx1536M` in `gradle.properties`.

I'm not sure what we can really do though if the problem is just memory usage. Might be that we need to claim that 16GB is the minimum specs on ChromeOS...

**Comment by andrewmk** [comment]
Hi. I've been hitting the same thing on my Chromebook. It doesn't (usually) happen if I'm just working from the command line but if I have Android Studio or VSCode open then it almost invariably gets stuck at the assembleDebug stage when doing a 'flutter run'. As commented by @lupohirp if the initial build works then hot reload or hot restart carry on working OK. I've got an Acer Chromebook CP713-1WN with an i5-8250U processor and 8GB ram, plenty of free disk space. When it happens the whole Linux VM side of things gets locked up - I can't open any more terminals or other Linux applications. Sometimes it recovers on its own, sometimes I have to restart the Chromebook. I'm happy to help trying to diagnose this. I swear it didn't happen when I first started playing with flutter on the Chromebook last year but going back to it a couple of weeks ago it happens a lot so maybe somethings changed (not helpful I know - it could be ChromeOS, could be any number of things).

**Comment by lupohirp** [comment]
> Hi. I've been hitting the same thing on my Chromebook. It doesn't (usually) happen if I'm just working from the command line but if I have Android Studio or VSCode open then it almost invariably gets stuck at the assembleDebug stage when doing a 'flutter run'. As commented by @lupohirp if the initial build works then hot reload or hot restart carry on working OK. I've got an [Acer Chromebook CP713-1WN](https://www.acer.com/ac/en/GB/content/professional-model/NX.EFJEK.003) with an i5-8250U processor and 8GB ram, plenty of free disk space. When it happens the whole Linux VM side of things gets locked up - I can't open any more terminals or other Linux applications. Sometimes it recovers on its own, sometimes I have to restart the Chromebook. I'm happy to help trying to diagnose this. I swear it didn't happen when I first started playing with flutter on the Chromebook last year but going back to it a couple of weeks ago it happens a lot so maybe somethings changed (not helpful I know - it could be ChromeOS, could be any number of things).

yes, this is the same thing that i hitted. i've also opened a issue to crbug to get a swap partition on crostini, because we can't work like this. Swap needs to be implemented if we want to do develop-intensive tasks with chromebooks even with 16 GB Ram. Please if you can star this issue....it will be a great help. https://bugs.chromium.org/p/chromium/issues/detail?id=1193485&q=&can=4

**Comment by andrewmk** [comment]
> yes, this is the same thing that i hitted. i've also opened a issue to crbug to get a swap partition on crostini, because we can't work like this. Swap needs to be implemented if we want to do develop-intensive tasks with chromebooks even with 16 GB Ram. Please if you can star this issue....it will be a great help. https://bugs.chromium.org/p/chromium/issues/detail?id=1193485&q=&can=4

I've starred the bug over on chromium. I'm still happy to see if there's anything to be done on the flutter side to make it work.

**Comment by Hixie** [comment]
@andrewmk @lupohirp By any chance did either of you experiment with the Gradle heap settings mentioned above?

**Comment by andrewmk** [comment]
@Hixie Yes I dropped it to 800M. I've had several successful runs in VSCode but several lockups too. Early experimentation suggests if I kill all java processes before trying the build (e.g. the gradle daemon) then it's more likely to work but I've only tried it a few times.

**Comment by lupohirp** [comment]
> @andrewmk @lupohirp By any chance did either of you experiment with the Gradle heap settings mentioned above?

yeah, but with the same results.

**Comment by lupohirp** [comment]
@andrewmk @Hixie  my bug over chromium has assigned. Seems that they will implement it. Or i'm wrong?

**Comment by Hixie** [comment]
> I've had several successful runs in VSCode but several lockups too

Yeah that's what I was worried about.

> my bug over chromium has assigned. Seems that they will implement it. Or i'm wrong?

I'm not familiar with how Chromium handles issues, sorry.

**Comment by lupohirp** [comment]
@andrewmk can you test with these gradle.properties(with visual studio and JDK 11)?

org.gradle.jvmargs=-XX:+UseContainerSupport -XX:MaxRAMPercentage=15.0
android.useAndroidX=true
android.enableJetifier=true
org.gradle.daemon=false

i've obtained quite good results

**Comment by andrewmk** [comment]
> @andrewmk can you test with these gradle.properties(with visual studio and JDK 11)?
> 
> org.gradle.jvmargs=-XX:+UseContainerSupport -XX:MaxRAMPercentage=15.0
> android.useAndroidX=true
> android.enableJetifier=true
> org.gradle.daemon=false
> 
> i've obtained quite good results

I haven't got much time at the moment but I've just used those settings and successfully done assembleDebug from within VSCode five times in a row which would not happen previously. So it looks like you might have hit on something. Massive thanks!

Prior to trying this I was executing 'killall java; killall dart' at the command line in between runs which seemed to help but is a bit brutal.

**Comment by lupohirp** [comment]
> > @andrewmk can you test with these gradle.properties(with visual studio and JDK 11)?
> > org.gradle.jvmargs=-XX:+UseContainerSupport -XX:MaxRAMPercentage=15.0
> > android.useAndroidX=true
> > android.enableJetifier=true
> > org.gradle.daemon=false
> > i've obtained quite good results
> 
> I haven't got much time at the moment but I've just used those settings and successfully done assembleDebug from within VSCode five times in a row which would not happen previously. So it looks like you might have hit on something. Massive thanks!
> 
> Prior to trying this I was executing 'killall java; killall dart' at the command line in between runs which seemed to help but is a bit brutal.

Great! I Must try also with Android Studio. Also Google ha set milestone on my opened bug: seems that they will implement in Chrome OS 93

**Comment by andrewmk** [comment]
I just tried Android Studio and got what I thought was a lock up on the second or third debug run but it eventually finished. Android Studio has always been a massive memory hog though and I'll be happy if I can use VSCode (though I am more familiar with Android Studio). Oddly when it happens it's not like it crashes with an out of memory error - it seems to max out the CPU(s) and the cooling fan goes crazy.

**Comment by wai** [comment]
> @andrewmk can you test with these gradle.properties(with visual studio and JDK 11)?
> 
> org.gradle.jvmargs=-XX:+UseContainerSupport -XX:MaxRAMPercentage=15.0
> android.useAndroidX=true
> android.enableJetifier=true
> org.gradle.daemon=false
> 
> i've obtained quite good results

Great! It saves my day! I confirm that make my android studio not stuck and work again! Thanks, @lupohirp  and @andrewmk.

I have the same machine model and spec with andrewmk (Acer Chromebook CP713-1WN with an i5-8250U processor and 8GB ram).

**Comment by Hixie** [comment]
This might be a good thing to document somewhere, or maybe we could detect low-memory systems and suggest it or some such.

**Comment by lupohirp** [comment]
> > @andrewmk can you test with these gradle.properties(with visual studio and JDK 11)?
> > org.gradle.jvmargs=-XX:+UseContainerSupport -XX:MaxRAMPercentage=15.0
> > android.useAndroidX=true
> > android.enableJetifier=true
> > org.gradle.daemon=false
> > i've obtained quite good results
> 
> Great! It saves my day! I confirm that make my android studio not stuck and work again! Thanks, @lupohirp and @andrewmk.
> 
> I have the same machine model and spec with andrewmk (Acer Chromebook CP713-1WN with an i5-8250U processor and 8GB ram).

how curious: we have the same spec (i5-8250U and 8GB Ram) and all of three have problems.

**Comment by lupohirp** [comment]
@andrewmk @wai just tuned JVM for balancing performance and Ram Usage. These are the best settings that i achieved: 

org.gradle.jvmargs=-XX:+UseContainerSupport -XX:MaxRAM=1024m -XX:MaxRAMPercentage=70.0
android.useAndroidX=true
android.enableJetifier=true
org.gradle.daemon=false

**Comment by andrewmk** [comment]
Just an update - this hasn't fixed the problem for me. I still get hangs in VSCode sometimes but at least it's rarer.

**Comment by dalewking** [comment]
I see the same that flutter is unusable from Android Studio, but works fine in VSCode on my brand new Chromebook Spin 713 with 11th gen i5 and 8GB.

**Comment by BElannaMichalke** [comment]
I am not entirely sure whether this is related, but my builds hang at an infinitely growing CPU load. After some seconds of build time, my load grows up to about 20 - 40 with all 8 CPU cores at 100 % while running on a quite powerful i7 CPU.

I'm running Chromium OS (self-built, based on FydeOS patches) on a ThinkPad Yoga 370 with 32 GiB of RAM.

**Comment by MrCsabaToth** [comment]
I have lowest config Pixelbook Go 8GB nem + 64GB SSD and I wasn't even able to build the first time. Tried Android studio and VSCode as well.
Since the Linux dev environment is a vm within crostini, I wonder if it's some kind of virtualization related race condition. I noticed that the Android studio file watcher plug-in craps itself (crashes too many times so studio yeets it).
Next time I'll try command line build.
I know 8 GB is not much but we should be able to build.

**Comment by MrCsabaToth** [comment]
I was running htop and it didn't look like a memory problem, about 4 gig was free out of 7 something in the vm. However something was burning the CPUs so hard the UI hardly responded, 30 second lags and stuff

**Comment by MrCsabaToth** [comment]
Further notices: the memory display of htop might be bogus. But the CPU bars are important, because they are colored. Just like with Sysinternal Process Explorer and other enhanced tools the red color means kernel time. When the build gets into a gridlock both CPUs are red, almost everything is in kernel space, something is deadlocked or spinning like crazy.

Fortunately the gradle.properties changes helped.

**Comment by MrCsabaToth** [comment]
I'm also trying [code-server](https://github.com/coder/code-server) which is a way to essentially run VS Code not inside the Linux VM as stand-alone electron app, but "outside" in Crostini's Chrome browser. Just now the build locked up again. I woke up the machine from sleep, not sure if that matters. According to htop as you can see memory is OK (although it looks it's even way too low occupancy, almost unrealistic?), and also the kernel time bound CPU utilization is very visible, the UI feedback was lagging like 30 sec or so:
![Screenshot 2022-08-06 9 24 11 PM](https://user-images.githubusercontent.com/744949/183278363-a1ceb94b-281f-4063-97fd-8ef4e6dc47f9.png)

**Comment by dalewking** [comment]
A year and a half later and still no progress on this issue?!?! I actually bought my Chromebook to do flutter development and this has made it a waste of money.

**Comment by MrCsabaToth** [comment]
In my case rarely the build is successful. I'm thinking what workaround could improve the success rate besides jvmargs tuning. For example building the same app and source on my beefy laptop, kicking it off, disconnecting the debugger, then plugging in the phone to the Chromebook. My use-case is mainly to debug my app out in the wild, and my main desktop replacement laptop is way less portable.
Is there a trick besides code-server to decrease memory pressure on the Linux VM?

**Comment by Pfredd** [comment]
I have an Asus CX9 Chromebook with an 11th gen I5 and 16GB RAM.

The title of this issue needs to be changed, as I had the same issue running in Android Studio and I have 16GB. Changing the gradle.properties seems to fix the issue, but I haven't tested it extensively.

Would it be possible to change the template for new Flutter projects so it uses the above mention gradle.properties?

**Comment by MrCsabaToth** [comment]
> I have an Asus CX9 Chromebook with an 11th gen I5 and 16GB RAM.
> 
> The title of this issue needs to be changed, as I had the same issue running in Android Studio and I have 16GB. Changing the gradle.properties seems to fix the issue, but I haven't tested it extensively.
> 
> Would it be possible to change the template for new Flutter projects so it uses the above mention gradle.properties?

How much memory do you configure for the Linux VM? I suspect that might be low

**Comment by Pfredd** [comment]
> 
> How much memory do you configure for the Linux VM? I suspect that might be low

Since this is running under ChromeOS's Linux subsystem, I don't have any control on how much RAM is available.

That being said, here is the output from the free command:

              total        used        free      shared  buff/cache   available
Mem:        14842824       94332    14254812        2376      493680    14748492
Swap:              0           0           0

So it looks like the VM thinks it has all memory...

**Comment by MrCsabaToth** [comment]
> > How much memory do you configure for the Linux VM? I suspect that might be low
> 
> Since this is running under ChromeOS's Linux subsystem, I don't have any control on how much RAM is available.
> 
> That being said, here is the output from the free command:
> 
> ```
> free
>                total        used        free      shared  buff/cache   available
> Mem:        14842824       94332    14254812        2376      493680    14748492
> Swap:              0           0           0
> ```
> 
> So it looks like the VM thinks it has all memory...

Sorry, I've sold my Chromebooks because of the memory issues, so I couldn't check what settings are there about this. (I had a 8GB and a 4GB Chromebook, and if I graduate to 16 GB laptop I'd buy a full fledged Linux laptop.)
1. However I have an advice [Matt Kaufman](https://twitter.com/the1mattkaufman) advised me - in case you are into Visual Studio Code: try [code-server](https://github.com/coder/code-server). This would at least take off a lot of the IDE memory usage from the Linux VM to the main Chrome OS Crostini side.
2. Upon reading about memory usage I've found this: https://www.aboutchromebooks.com/news/chrome-os-update-will-allow-higher-end-chromebooks-to-use-more-memory-for-linux/ . So supposedly in 2019 there were commits already about letting the Linux VM dynamically cannibalize more memory. So maybe the issue is not just about memory, maybe there's something Chrome OS specific?

**Comment by maks** [comment]
I'll chime in here to say that I got here while searching issues due to experiencing similar symptoms on my Pixel Chromebook with 8GB ram. BUT in my case it's not anything to do with the JVM as I am using VSCode and building a debug build (`flutter run`) my Flutter app for Linux desktop, **not** Android.

The symptom I experience is as described in previous comments: CPU usage goes very high inside the Linux VM to the point where bash shells become barely responsive, both htop and Cog (chrome extension to show system stats) shows memory is still available and most importantly the rest of ChromeOS remains responsive. 

I'll try to run some tracing to see whats happening in my casem but I wonder if it's that not enough CPU is being allocated to the Linux VM?

**Comment by lupohirp** [comment]
> 
> I'll try to run some tracing to see whats happening in my casem but I wonder if it's that not enough CPU is being allocated to the Linux VM?

That could be. Infact https://chromium-review.googlesource.com/c/crosvm/crosvm/+/4208668 they are working on this aspect.

**Comment by MrCsabaToth** [comment]
> > I'll try to run some tracing to see whats happening in my casem but I wonder if it's that not enough CPU is being allocated to the Linux VM?
> 
> That could be. Infact https://chromium-review.googlesource.com/c/crosvm/crosvm/+/4208668 they are working on this aspect.

What do you mean by not enough CPU? Computation power should not matter, it should just result in longer compilation, unless there's some time dependent check which could trip. Sometimes I was able to compile, so it didn't fail all the time. That points rather to some kind of a race condition induced by low memory to me. I moved away from Chromebook because of these issues.
